[WIP] initial implementation

AdaskoTheBeAsT opened the pull request:
Work in progress - this is for gathering feedback

nils-a reviewed (changes requested):
I like it. 👍 

There are some things I would change:

### `CakeNamespaceImport`

see https://cakebuild.net/docs/extending/addins/creating-addins#importing-namespaces

### Namespace documentation

see https://cakebuild.net/community/contributing/documentation#namespaces

### Collisions

Currently, the addin works fine when run from Cake frosting. However running from Cake script runner creates the following warning:

```text
Namespace Cake.ReSharper.GlobalTools.InspectCode excluded by code generation, affected methods:
        method "InspectCode" excluded from code generation and will need to be fully qualified on ICakeContext.
        method "InspectCode" was included in code generation, but will need to be fully qualified on ICakeContext.
        method "InspectCodeFromConfig" excluded from code generation and will need to be fully qualified on ICakeContex
```

My first guess here is that all three methods exist with exactly the same name in `Cake.Common.Tools.InspectCode`. This leads to exactly what the warning says: `InspectCode(string)` does not use the code from this addin but rather `Cake.Common.Tools.InspectCode.InspectCode(string)`

Not sure what a good naming would be, but having collisions with other classes leads to "bad UX".

AdaskoTheBeAsT commented:
@nils-a thanks a lot for feedback! I will adjust code accordingly.
Based on observations of the whole code base what you think about 
prefixing method names just like for example DotNetBuild
so InspectCode -> ReSharperInspectCode
CleanupCode -> ReSharperCleanupCode?

nils-a commented:
> `ReSharperInspectCode`, `ReSharperCleanupCode`

Probably. Maybe. I'm not really fond of that. I have no better idea, though. 😁 

On the other hand - that's what it is, right?
Let's see if @augustoproiete or @gep13 have some ideas.

AdaskoTheBeAsT commented:
hi @nils-a could you please take a look? I added NamespaceDoc, changed names to not cause conflicts and also added namespace imports (not sure if tool namespace import is relevant)

nils-a commented:
@AdaskoTheBeAsT love it. One problem is, I'm seeing an error at the end of the run:

```text
An error occurred when executing task 'Default'.
Error: One or more errors occurred. (Access to the path 'C:\_dev\out.json' is denied.)
        Access to the path 'C:\_dev\out.json' is denied.
```

Where `C:\_dev\out.json` is my `OutputFile`. I'm guessing this is in `AnalyzeResultsFile` where the file is opened. My hunch would be that ReSharper tools are not "done" writing/closing the file when you try to read it.

nils-a commented:
Also, looking at `AnalyzeResultsFile` I get the impression that it's always reading XML. Does this still work when I set `OutputFileFormat = ReSharperInspectCodeReportFormat.Json` in `ReSharperInspectCodeSettings` ?

AdaskoTheBeAsT commented:
@nils-a you are absolutely right.
I was concentrated on mecahnical "stick with cake code style" that I skipped thinking about consequences of settings

I plan to provide IReSharperInspectCodeResultAnalyzer with implementation for every report format and factory which will return proper result analyzer based on report format.
Also one more question - what is your routine of testing such plugins to cake? do you pack them in nuget and host locally with nuget config additional source pointing to folder or there is more efficient way of testing them?
What I plan is to test them also on WSL2 and MacOS.

AdaskoTheBeAsT commented:
@nils-a in last commit initial extract how it can look like based on xml sample

AdaskoTheBeAsT commented:
...but trick will be with html report
of course I can use https://html-agility-pack.net/from-file

nils-a commented:
> what is your routine of testing such plugins to cake?

You mean integration tests? 

Usually, I have a `demo` folder in the project which contains the demo (and integration tests). In there I reference the dll directly: For Cake scripts this is done via the [`#reference` preprocessor directive](https://cakebuild.net/docs/writing-builds/preprocessor-directives/reference), and for Cake Frosting simply by adding a reference in the csproj.

This place then is a bit of mix-and-mingle for me: There are "integration tests", demos and probably also some issue-related tests. You can look at https://github.com/cake-contrib/Cake.7zip/blob/develop/demo/ to see how I did that. However, the "integration tests" are more like run-and-see-the-outcome, so not real tests.

If you need a nice example of real integration tests, then there is [Cake.Git](https://github.com/cake-contrib/Cake_Git/blob/8599fc28aa76f34d25b6b5ed6048949a90751321/test.cake). Essentially doing something, then verify the expected outcome and throw some `Exception` on failure.

AdaskoTheBeAsT commented:
I see - ok next commit will contain integration tests - let me take a look and prepare such things - probably it will take me few days

nils-a commented:
> but trick will be with html report

Interesting problem. I guess it's not possible to write two reports at once, so one can always be an xml report? 😁

AdaskoTheBeAsT commented:
@nils-a no it is not possible - anyway for html report I dropped initial idea of using HtmlAgilityPack - simple regex will do the trick

unfortunately html report do not contains severity of issues (limitiation of the tool)
also in case of json report - it is not one file generated but folder with jsons

I added initial integ tests based on great samples - I will continue work 
finishing for today - 4:15am - time to rest

AdaskoTheBeAsT commented:
Frosting version works on WSL2 Ubuntu-20.4 after installing libgdiplus libc6-dev
```
sudo apt install libgdiplus libc6-dev
```
now I am checking cake version

![image](https://user-images.githubusercontent.com/1914201/153765505-17496dca-e5a9-4fb8-a63f-b373222382f8.png)

AdaskoTheBeAsT commented:
ok some small errors in casing
now it works on linux 

I will take a look how to normalize output lines for different formats and tomorrow 
I will check how it works on macos


![image](https://user-images.githubusercontent.com/1914201/153765900-e5b88bed-18a6-40b1-8bd9-a18477995731.png)

AdaskoTheBeAsT commented:
from MacOS cake integ test
<img width="1204" alt="image" src="https://user-images.githubusercontent.com/1914201/153827364-07e1b467-95fb-44b9-8b24-87c970f60e89.png">

AdaskoTheBeAsT commented:
MacOS
frosting
<img width="1155" alt="image" src="https://user-images.githubusercontent.com/1914201/153827607-bce403f8-bdf3-4da0-8af4-0ddaccaa3263.png">

AdaskoTheBeAsT commented:
> @AdaskoTheBeAsT love it. One problem is, I'm seeing an error at the end of the run:
> 
> ```
> An error occurred when executing task 'Default'.
> Error: One or more errors occurred. (Access to the path 'C:\_dev\out.json' is denied.)
>         Access to the path 'C:\_dev\out.json' is denied.
> ```
> 
> Where `C:\_dev\out.json` is my `OutputFile`. I'm guessing this is in `AnalyzeResultsFile` where the file is opened. My hunch would be that ReSharper tools are not "done" writing/closing the file when you try to read it.


@nils-a it seems that ReSharper tools instead of creating one file - create folder with multiple files - it was fixed in analyzer - please recheck

nils-a commented:
@AdaskoTheBeAsT I like it. (Had some trouble running the integration tests on windows - but that might well be my system. Everything was fine on ubuntu.)

AdaskoTheBeAsT commented:
@nils-a majority of the code is original Cake implementation :) it is good by design 
I provided only small adjustments ;)
one more thing - building pipeline - should use original cake repository for crafting build pipeline?
what is general direction of cake building process? it is github actions invoking cake build script or azure devops invoking cake build script?

nils-a commented:
> what is general direction of cake building process? it is github actions invoking cake build script or azure devops invoking cake build script?

That's totally up to you. Traditionally, we've had many builds running on AppVeyor. I, personally, tend to switch all "my" builds to GitHub actions. If I remember correctly, @pascalberger prefers DevOps. So, whatever fits you best.

> should use original cake repository for crafting build pipeline?

This is also totally up to you. The Cake pipeline is fairly complex and does many things, so I probably would not start out by "copying" that. I, personally, prefer using [Cake.Recipe](https://github.com/cake-contrib/Cake.Recipe) which makes many things real easy - but it also forces you to think a bit in a Cake.Recipe-way (and for the moment, it still uses Cake 0.38.5). One example of using Cake.Recipe is [Cake.7zip](https://github.com/cake-contrib/Cake.7zip/blob/develop/recipe.cake). An example for a hand-craftet pipeline is [Cake.Git](https://github.com/cake-contrib/Cake_Git/blob/develop/build.cake). But again, chose whatever fits you best.

AdaskoTheBeAsT commented:
@nils-a for now I started with Azure DevOps as it is most familiar to me - I will try github actions later - how to best generate documentation from code?

nils-a commented:
@AdaskoTheBeAsT The [Guidelines](https://github.com/cake-contrib/Home/blob/master/Guidelines.md#addin-documentation) currently suggest using [wyam](https://wyam.io), however, I am assuming this will change in the future to suggesting [statiq](https://www.statiq.dev). In the end (as with most things) this is up to you - what you feel most comfortable with.

If you push the documentation to a [branch called `gh-pages`](https://www.gep13.co.uk/blog/how-to-create-gh-pages-branch) documentation will be automatically available via GitHub pages under https://cake-contrib.github.io/Cake.ReSharper.GlobalTools/. 

You can have a look at https://cake-contrib.github.io/Cake.7zip/ - which is generated from https://github.com/cake-contrib/Cake.7zip/tree/develop/docs/ by the `Publish-Documentation` task of [Cake.Recipe](https://github.com/cake-contrib/Cake.Recipe/blob/develop/Source/Cake.Recipe/Content/wyam.cake#L11)

AdaskoTheBeAsT commented:
@nils-a thanks - I will take a look
